processing/src/main/java/io/druid/query/groupby/epinephelinae/ConcurrentGrouper.java

```diff
@@ -292,19 +325,413 @@ private boolean isParallelSortAvailable()
+      final List<String> dictionary = grouper.getDictionary();
+      totalDictionarySize += dictionary.stream().mapToLong(RowBasedGrouperHelper::estimateStringKeySize).sum();
+      if (totalDictionarySize > maxDictionarySize) {
+        return null;
+      }
+      mergedDictionary.addAll(dictionary);
+    }
+
+    return ImmutableList.copyOf(mergedDictionary);
+  }
+
+  /**
+   * Build a combining tree for the input iterators which combine input entries asynchronously.  This method is called
+   * when data are spilled and thus streaming combine is preferred to avoid too many disk accesses.
+   *
+   * @param sortedIterators sorted iterators
+   *
+   * @return an iterator of the root grouper of the combining tree
+   */
+  private Iterator<Entry<KeyType>> parallelCombine(
+      List<Iterator<Entry<KeyType>>> sortedIterators,
+      List<String> mergedDictionary
+  )
+  {
+    // CombineBuffer is initialized when this method is called
+    final ResourceHolder<ByteBuffer> combineBufferHolder = combineBufferSupplier.get();
+    final ByteBuffer combineBuffer = combineBufferHolder.get();
+    final AggregatorFactory[] combiningFactories = new AggregatorFactory[aggregatorFactories.length];
+    for (int i = 0; i < aggregatorFactories.length; i++) {
+      combiningFactories[i] = aggregatorFactories[i].getCombiningFactory();
+    }
+    final int minimumRequiredBufferCapacity = StreamingMergeSortedGrouper.requiredBufferCapacity(
+        combineKeySerdeFactory.factorizeWithDictionary(mergedDictionary),
+        combiningFactories
+    );
+    final Pair<Integer, Integer> degreeAndBufferNum = findCombineDegreeAndNumBuffers(
+        combineBuffer,
+        minimumRequiredBufferCapacity,
+        concurrencyHint,
+        sortedIterators.size()
+    );
+
+    if (degreeAndBufferNum == null) {
+      throw new ISE("Cannot find a proper combine degree");
+    }
+
+    final int combineDegree = degreeAndBufferNum.lhs;
+    final int bufferNum = degreeAndBufferNum.rhs;
+    final int sliceSize = combineBuffer.capacity() / bufferNum;
+
+    final List<ByteBuffer> slices = new ArrayList<>(bufferNum);
+    for (int i = 0; i < bufferNum; i++) {
+      slices.add(getSlice(combineBuffer, sliceSize, i));
+    }
+
+    final Supplier<ByteBuffer> bufferSupplier = new Supplier<ByteBuffer>()
+    {
+      private int i = 0;
+
+      @Override
+      public ByteBuffer get()
+      {
+        if (i < slices.size()) {
+          return slices.get(i++);
+        } else {
+          throw new ISE("Requested number of buffer slices exceeds the planned one");
+        }
+      }
+    };
+
+    final List<Future> combineFutures = new ArrayList<>(bufferNum);
+    final Iterator<Entry<KeyType>> combinedIterator = buildCombineTree(
+        sortedIterators,
+        bufferSupplier,
+        combiningFactories,
+        combineDegree,
+        combineFutures,
+        mergedDictionary
+    );
+
+    return new Iterator<Entry<KeyType>>()
```

Comment:
This may be refactored as generic delegating iterator "with baggage to close". Also consider implementing `CloseableIterator` instead of relying on `hasNext()`.

Reply:
Done.